### DISCUSSION TITLE
initialize multiply times when enable ANYWIDGET_HMR

### DISCUSSION THREAD

**Discussion by Roy-Kid** [manzt/anywidget#683] [General] [answered]
Hi, I found a strange behavior when I am developing.
I am using 
```
%load_ext autoreload
%autoreload 2
%env ANYWIDGET_HMR=1
import molvis
```
and `vite dev` and `_DEV`-style for developing, and when I re-initialize the widget class, both python class and backend will initialize multiply time. For example, the first time I open vscode and initialize it, the class will initialize once, and re-initialize it, it will initialize twice, and so on. Even after I restart the kernel

Here is the log:
![image](https://github.com/user-attachments/assets/df2f69fe-4a01-4114-93c8-ab2e70ca8fb4)
after a cell run, the class initialize multiply times(it's very regular, the n-th run will cause n times initialization). 

Any good idea to fix it? or maybe I can not use autoreload with ANYWIDGET_HMR?

**Comment by Roy-Kid** [comment]
also, even though I call a method of widget, let's say "molvis.add_atom()", the widget also will be re-initialized

**Reply by manzt** [reply, marked as answer]
Hi there, sorry for the delay. I think this could be due to mixing anywidget's HMR with Vite. If you are using Vite, I would not set the `ANYWIDGET_HMR=1` since vite will take care of hot-reloading your JS modules. I'm thinking that what is happening right now is that both anywidget and vite are trying to refresh at the same time.

**Reply by Roy-Kid** [reply]
Thanks! I think I mixed this refresh problem with another one: when I re-execute the cell, the render function will also be called. I re-wrote the code and this double refresh issue is obvious. I already fixed it! Again, thanks a lot!
